Stop Viewer Departure Layout

amy-corson-ibigroup opened the pull request:
Separates upcoming stop times by day and tweaks flexbox/grid elements to avoid clipping and overflow at smaller screen sizes.

Definitely looking for feedback around how the times are being sorted by day (it feels like there may be a more efficient way to render each day!)

Before                                 |  After
:-------------------------:|:-------------------------:
![image](https://user-images.githubusercontent.com/115499534/217331841-960d64b1-558b-4962-b8ac-0c3aad149219.png)  |  
![image](https://user-images.githubusercontent.com/115499534/217333085-93834f6b-9492-478d-8111-de97c2d3fc0a.png)

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -151,35 +159,143 @@ class LiveStopTimes extends Component<Props, State> {
-                  stopTimes={times}
-                  stopViewerArriving={stopViewerArriving}
-                  stopViewerConfig={stopViewerConfig}
+        <div>
```
Do we need this outermost div?

miles-grant-ibigroup commented on lib/components/viewers/pattern-row.tsx:
```diff
@@ -73,31 +73,37 @@ class PatternRow extends Component<Props, State> {
 
     const routeName = route.shortName ? route.shortName : route.longName
     const routeColor = getRouteColorBasedOnSettings(route.operator, route)
+    console.log(stopTimes)
```
Whoops I think this slipped in!

miles-grant-ibigroup commented on lib/components/viewers/pattern-row.tsx:
```diff
@@ -73,31 +73,37 @@ class PatternRow extends Component<Props, State> {
+    console.log(stopTimes)
+    let style
+    if (routeName && routeName?.length >= 4 && routeName?.length <= 6) {
+      style = { fontSize: '20px' }
```
Instead of setting the entire style object, we should instead set a font size variable we insert later

miles-grant-ibigroup commented on lib/components/viewers/pattern-row.tsx:
```diff
@@ -73,31 +73,37 @@ class PatternRow extends Component<Props, State> {
     const routeColor = getRouteColorBasedOnSettings(route.operator, route)
+    console.log(stopTimes)
+    let style
+    if (routeName && routeName?.length >= 4 && routeName?.length <= 6) {
```
I don't think you need the first routeName check! Since you have the question mark

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,71 @@
   font-size: 42px;
   gap: 5px;
-  grid-template-columns: 4ch 2fr;
+  grid-template-columns: fit-content(8ch) fit-content(8ch);
```
This causes the grid to sometimes flow strangely, with destinations being unaligned. I think we need a better solution here

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,71 @@
-  grid-template-columns: 4ch 2fr;
+  grid-template-columns: fit-content(8ch) fit-content(8ch);
 }
 .otp .stop-viewer .route-row .header .route-name span {
```
I'm not a fan of the webkit specific rules... What's going on here?

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -246,7 +286,7 @@
   align-items: center;
   display: flex;
-  max-width: 250px;
+  width: 75%;
```
This might also be what's causing the grid misalignment?

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -151,35 +159,143 @@ class LiveStopTimes extends Component<Props, State> {
+                {todayTimes.map(({ id, pattern, route, times }) => {
+                  // Only add pattern if route info is returned by OTP.
+                  return (
+                    <PatternRow
```
I don't like repeating the same code 3 times. Is there any way to replace the 3 arrays with a new field inside of a new single array that's date or something, we can then check for that date changing inside of the map and render a heading accordingly. Let me know if you'd like to discuss this more offline, I'm not sure if that was super clear

miles-grant-ibigroup reviewed (changes requested):
The overall vision is good! Just a few specifics that I think we need to figure out

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +150,88 @@ class LiveStopTimes extends Component<Props, State> {
+      )
+
+    const routesByDay: any = []
+    const firstDayOfService = routeTimes[0].times[0].serviceDay
```
```suggestion
    const firstDayOfService = routeTimes?.[0].times?.[0].serviceDay
```

We can't rely on this existing!
<img width="612" alt="image" src="https://user-images.githubusercontent.com/86619099/220986794-ebc7b405-a16e-4c80-9b79-0f6b1ae98ec8.png">

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +150,88 @@ class LiveStopTimes extends Component<Props, State> {
+    const firstDayOfService = routeTimes[0].times[0].serviceDay
+
+    // Loop through all routes, and sort based on that route's next available day of service.
+    routeTimes.forEach((route) => {
+      const { serviceDay } = route.times[0]
+      const daysPastFirstService =
+        (serviceDay - firstDayOfService) / ONE_DAY_IN_SECONDS
+
+      // Create a separate array for each service day
+      routesByDay[daysPastFirstService]
+        ? routesByDay[daysPastFirstService].push(route)
+        : routesByDay.push([route])
+    })
```
I'm not sure I'm crazy about `routesByDay` wrapper array. Would it not be possible to just add a `day` field to the route time object?

amy-corson-ibigroup replied:
Added in [e6abad9](https://github.com/opentripplanner/otp-react-redux/pull/780/commits/e6abad93592a98933c323342b67150019b12ff32)!

miles-grant-ibigroup reviewed (changes requested):
I think we're getting there! Just a few small architecture changes -- I think we can get it cleaner

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +148,81 @@ class LiveStopTimes extends Component<Props, State> {
+      )
+
+    // Next 3 service days at this stop
+    const daysOfService = [
+      ...Array.from(new Set(routeTimes.map((route) => route.day)))
+    ]
+      .sort(function (a, b) {
+        return a - b
+      })
+      .slice(0, 3)
```
I think we can get away with not needing this! See #806

miles-grant-ibigroup commented on lib/components/viewers/pattern-row.tsx:
```diff
@@ -73,31 +70,36 @@ class PatternRow extends Component<Props, State> {
 
     const routeName = route.shortName ? route.shortName : route.longName
     const routeColor = getRouteColorBasedOnSettings(route.operator, route)
+    let fontSize = '32px'
+    if (routeName) {
+      if (routeName?.length >= 4 && routeName?.length <= 6) {
+        fontSize = '20px'
+      } else if (routeName?.length > 7) {
+        fontSize = '16px'
+      }
+    }
```
Can this be replaced with `routeNameFontSize`?

amy-corson-ibigroup replied:
Yes good catch!!

miles-grant-ibigroup reviewed (changes requested):
Getting there! Just a little bit more cleanup

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,86 @@
+  gap: 10px;
 }
 .otp .stop-viewer .route-row .header .route-name span {
+  display: -webkit-box;
   font-size: 15px;
+  line-clamp: 3;
+  overflow: hidden;
+  -webkit-box-orient: vertical;
+  -webkit-line-clamp: 3;
```
I'm not sure we should use webkit specific rules. We have a post processor that's supposed to automatically apply these if we use the generic rules! What is this doing anyway? When we originally built this we moved away from `line-clamp` because it was so poorly supported

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,86 @@
+  -webkit-box-orient: vertical;
+  -webkit-line-clamp: 3;
 }
 .otp .stop-viewer .route-row .header .route-name strong span {
```
Some of these rules could probably be replaced by a stronger grid for the entire row, but this isn't blocking.

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,86 @@
 
 /* stop viewer styles */
 
-.otp .stop-viewer .route-row {
```
Any way we can set some form of min-height on this? To prevent this issue: 
![image](https://user-images.githubusercontent.com/86619099/223164147-9e9ca983-461b-4a7e-87a0-55914c6aa677.png)

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +169,72 @@ class LiveStopTimes extends Component<Props, State> {
+          // Only show times within 24 hours of next arrival time
+          .filter((time: any, i: number, array: Array<any>) => {
+            const firstDepartureTime =
+              array[0].serviceDay + array[0].realtimeDeparture
```
```suggestion
              array?.[0]?.serviceDay + array?.[0]?.realtimeDeparture
```

We can't rely on any of this data!

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +169,72 @@ class LiveStopTimes extends Component<Props, State> {
+            const departureTime = time.serviceDay + time.realtimeDeparture
+            return i === 0 || (departureTime - firstDepartureTime) / 3600 < 24
+          })
+        const { serviceDay } = sortedTimes[0]
```
I see this causing crashes too. Maybe we can add some sort of null check?

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +169,72 @@ class LiveStopTimes extends Component<Props, State> {
+        }
+      })
+      .filter(({ pattern, route }) =>
+        routeIsValid(route, getRouteIdForPattern(pattern))
```
Any way we can move that previous filter into this filter step? Just so we don't loop through the array twice

amy-corson-ibigroup replied:
I can move the `routeIsValid` filter into the `times !== 0` filter no problem! The 2nd service day filter below is a bit tricker, because it relies on the `sort(stopTimeComparator)` which  in turn needs the `times !== 0` filter to avoid throwing an error. There could definitely be a better way to do this that I'm missing!

miles-grant-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -144,36 +169,72 @@ class LiveStopTimes extends Component<Props, State> {
+        routeIsValid(route, getRouteIdForPattern(pattern))
+      )
+      .filter((route) => {
+        /* If the route's first departure time falls on the 2nd available service day,
+      only show it if it's within 6 hours of now. */
+        if (!isSameDay(utcToZonedTime(route.day * 1000, homeTimezone), now)) {
+          const departureTime = route.times[0].realtimeDeparture + route.day
+          return isBefore(departureTime * 1000, addHours(now, 6))
+        }
+        return route
```
And ideally combine this one as well. There's some shared logic here as well so I think we can get clever with it

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,86 @@
+  overflow: hidden;
+  padding-left: 15px;
+  text-align: center;
+  text-overflow: ellipsis;
```
This sometimes causes strange results... 
<img width="435" alt="Screenshot 2023-03-06 at 11 12 17 AM" src="https://user-images.githubusercontent.com/86619099/223167025-eb6529ff-b1af-44fd-92a7-25f6490eb711.png">

miles-grant-ibigroup reviewed (changes requested):
This PR is such a tricky one but I think we're getting there. Let me know if you'd like to chat about this offline a bit there's some open questions in the air

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,85 @@
+
+.otp .stop-viewer .day-container .route-row-container li .header.stop-view {
+  border-radius: 0;
+  height: fit-content;
```
This is still creating varied row heights for me!

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -246,7 +301,7 @@
   align-items: center;
   display: flex;
-  max-width: 250px;
+  max-width: 75%;
```
Percentages on text can be wonky especially on strange screen widths. Can we do this with grid instead?

miles-grant-ibigroup commented on lib/components/viewers/pattern-row.tsx:
```diff
@@ -111,7 +97,7 @@ class PatternRow extends Component<Props, State> {
               </div>
             </strong>
-            <span>{pattern.headsign}</span>
+            <span title={pattern.headsign}>{pattern.headsign}</span>
```
I forget if we discussed this already but I don't think it's good practice to have the `title` contain the exact same content as the element itself!

amy-corson-ibigroup replied:
In this case I think it's useful when often we're seeing gtfs where the shortname or headsign significantly overflows the container! A stopgap so users still have access to the info

miles-grant-ibigroup reviewed (changes requested):
Looking much much cleaner! However still seeing some potential css issues, plus we're getting a lot of useless divs in the markup.

There are also merge conflicts

miles-grant-ibigroup commented on lib/components/viewers/viewers.css:
```diff
@@ -102,48 +102,76 @@
   list-style-type: none;
   padding-left: 0;
 }
+.otp .stop-viewer .departures {
```
Can we add this as a comma rule for the related-panel-container

miles-grant-ibigroup reviewed (approved):
Approved pending the changes we discussed. good luck with the merge conflicts!

miles-grant-ibigroup commented:
A little worried about that snapshot update commit -- it's updating a lot of things that this PR wasn't supposed to touch?

amy-corson-ibigroup commented:
I saw that, I poked around for a while last night and couldn't figure out why! I figured I'd wait to investigate in case the next round of review requires another snapshot update, but I can revert the commit if that would be better.

binh-dam-ibigroup reviewed (changes requested):
Implement the change we agreed upon during the team discussion to reuse/repurpose the config item for how long in advance to display upcoming trips. Otherwise, looks solid!

amy-corson-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -155,36 +180,75 @@ class LiveStopTimes extends Component<Props, State> {
-            )
-            .map(({ id, pattern, route, times }) => {
-              // Only add pattern if route info is returned by OTP.
+        <div className="departures">
```
Good catch thank you!

amy-corson-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -155,36 +180,75 @@ class LiveStopTimes extends Component<Props, State> {
+      })
+      .filter((route) => {
+        /* If the route's first departure time falls on the 2nd available service day,
+      only show it if it's within 6 hours of now. */
```
Yes, however the changes we discussed on the tech call removed this filter & comment entirely!

binh-dam-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -124,6 +128,26 @@ class LiveStopTimes extends Component<Props, State> {
+    const formattedDay = utcToZonedTime(day * 1000, homeTimezone)
+    return (
+      <React.Fragment key={day}>
+        {/* If the service day is not today, add a label */}
```
A very small thing... inserting a label visually "breaks" the grouping (the edges of the arrivals it sit between are not rounded).

amy-corson-ibigroup replied:
This is actually intentional! The idea was for the entire departure container to be rounded, so the middle rows should stay square. Original wireframe for reference: 

![image](https://user-images.githubusercontent.com/115499534/227360197-39ab89ad-3d53-459b-917c-837240c24dd7.png)

binh-dam-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -151,25 +176,55 @@ class LiveStopTimes extends Component<Props, State> {
       return stopTimeComparator(stopTimesA[0], stopTimesB[0])
     }
 
+    const configDepartureWindow = stopViewerConfig?.timeRange / 86400
```
Per how the redux state is initialized, `stopViewerConfig.timeRange` should always be defined.

binh-dam-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -151,25 +176,55 @@ class LiveStopTimes extends Component<Props, State> {
+        const sortedTimes = route.times
+          .concat()
+          ?.sort(stopTimeComparator)
+          // Only show times within the next 2 days, unless otherwise indicated by the config.
```
Update the redux initial state too. The default there is still 4 days, see https://github.com/opentripplanner/otp-react-redux/blob/46c32ead25dc421d7baba0c4df9a2097ea2c6961/lib/reducers/create-otp-reducer.js#L87.

binh-dam-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -151,25 +176,55 @@ class LiveStopTimes extends Component<Props, State> {
+          ?.sort(stopTimeComparator)
+          // Only show times within the next 2 days, unless otherwise indicated by the config.
+          .filter((time: any, i: number, array: Array<any>) => {
+            const departureWindow = configDepartureWindow || 4
```
Isn't this variable redundant now? (`configDepartureWindow` always defined -- assuming it is probably non-zero?)

binh-dam-ibigroup commented on lib/components/viewers/live-stop-times.tsx:
```diff
@@ -151,25 +176,55 @@ class LiveStopTimes extends Component<Props, State> {
+                    time.day * 1000,
+                    routeTimes[index - 1]?.day * 1000
+                  )) ||
+                  // TODO: is new Date() the right approach?
```
Delete this comment.

binh-dam-ibigroup reviewed (approved):
There is one missing change to how the default time range is initialized in the redux state, please fix that, otherwise changes look good.

amy-corson-ibigroup commented:
The a11y tests were failing because technically the `renderDay` label is a `<p>` sitting within a `<ul>`. To work around this, I added `aria-hidden` to the `p` and added an `InvisibleA11yLabel` to each time that does not fall on the current day.

| UI    | Screenreader |
| -------- | ------- |
|![image](https://user-images.githubusercontent.com/115499534/227362062-1779cf15-4144-46d7-837b-c33b41a44a9d.png)|![image](https://user-images.githubusercontent.com/115499534/227362284-54d1ca31-452c-4bec-9e2d-a9e476253bf6.png)|

Tried using aria-role, but it kept failing! Let me know if this doesn't seem like the best solution.

amy-corson-ibigroup commented:
A poorly executed merge from dev ([2aa1792](https://github.com/opentripplanner/otp-react-redux/pull/780/commits/2aa179263ddf8605729baee390258640da1bdc3c)) could cause potentially hundreds of merge conflicts in weblate, so closing in favor of #837!
